src/librustc_parse/parser/mod.rs

```diff
@@ -1066,39 +1088,6 @@ impl<'a> Parser<'a> {
         }
     }
 
-    pub fn process_potential_macro_variable(&mut self) {
-        let normalized_token = match self.token.kind {
-            token::Dollar
-                if self.token.span.from_expansion() && self.look_ahead(1, |t| t.is_ident()) =>
```

Comment:
Oh man, this `from_expansion`-based check is totally incorrect.

If we take passing test https://github.com/rust-lang/rust/blob/master/src/test/ui/issues/issue-40847.rs and imagine that `gen!(m($var) $var);` is produced by a proc macro, it will incorrectly fail with an "unknown macro variable" error.

Reply:
We certainly do not want to check expansion kinds (was this token produced by `macro_rules` or not) from parser, so the only choice is to not check for `$ IDENT` ahead of time and delay it until the time when it's actually unexpected (i.e. no https://github.com/rust-lang/rust/pull/69211#discussion_r379913103).

Reply:
I think I'll just remove the special diagnostic support for `$ IDENT` from this PR.
I've filed https://github.com/rust-lang/rust/issues/69229 for a better fix.

Reply:
Hmm, OK -- I was about to r+. Let me know when it's ready for review again. :)